### PR TITLE
Clean test suite

### PR DESCRIPTION
**Pull request by ricardopereira**
#652

### CODE REVIEW

**Comment by ricardopereira** [comment]
@tcard @mattheworiordan Please review this when possible. I will merge this because I need to release a version and I'm confident about it but please open a issue if you noticed something wrong or suspicious.

**Comment by mattheworiordan** [comment]
@ricardopereira wow, lots of work there, will do my best to review. @tcard is unfortunately not at Ably anymore.

**Comment by mattheworiordan** [comment]
@funkyboy could you review this perhaps?

File: Source/ARTOSReachability.m [hunk 1]
```diff
@@ -86,7 +86,7 @@ - (void)off {
 - (void)internalCallback:(BOOL)reachable {
     [_logger info:@"Reachability: host %@: %d", _host, reachable];
     dispatch_async(_queue, ^{
-        _callback(reachable);
+        if (_callback) _callback(reachable);
```
Comment (funkyboy) [inline Source/ARTOSReachability.m]: I'd use this style here:

```
if (_callback) {
 _callback(reachable);
}
```

Or so the heartbleed bug taught me :)

File: Source/ARTQueuedMessage.m [hunk 2]
```diff
@@ -29,19 +33,30 @@ - (NSString *)description {
...
     return NO;
 }
 
-- (void (^)(ARTStatus *))cb {
-    return ^(ARTStatus * status) {
-        for (void (^cb)(ARTStatus *) in self.cbs) {
+- (void (^)(ARTErrorInfo *))sentCallback {
+    return ^(ARTErrorInfo *error) {
+        for (void (^cb)(ARTErrorInfo *) in self.sentCallbacks) {
+            cb(error);
```
Comment (funkyboy) [inline Source/ARTQueuedMessage.m]: I'd use `callback` instead of `cb` here

File: Source/ARTRealtime.m [hunk 3]
```diff
@@ -932,69 +951,77 @@ - (void)sendImpl:(ARTProtocolMessage *)msg callback:(void (^)(ARTStatus *))cb {
     NSData *data = [self.rest.defaultEncoder encodeProtocolMessage:msg error:&error];
 
     if (error) {
-        cb([ARTStatus state:ARTStateError info:[ARTErrorInfo createFromNSError:error]]);
+        ARTErrorInfo *e = [ARTErrorInfo createFromNSError:error];
+        if (sentCallback) sentCallback(e);
+        if (ackCallback) ackCallback([ARTStatus state:ARTStateError info:e]);
```
Comment (funkyboy) [inline Source/ARTRealtime.m]: Also here, in both cases, I'd use curly braces.

File: Source/ARTRealtime.m [hunk 4]
```diff
@@ -932,69 +951,77 @@ - (void)sendImpl:(ARTProtocolMessage *)msg callback:(void (^)(ARTStatus *))cb {
...
         }
         if (!merged) {
-            ARTQueuedMessage *qm = [[ARTQueuedMessage alloc] initWithProtocolMessage:msg callback:cb];
+            ARTQueuedMessage *qm = [[ARTQueuedMessage alloc] initWithProtocolMessage:msg sentCallback:nil ackCallback:ackCallback];
             [self.queuedMessages addObject:qm];
         }
-    } else {
+    }
+    else {
         // TODO review error code
```
Comment (funkyboy) [inline Source/ARTRealtime.m]: Is this still needed? If yes, I'd open an issue.

**Comment by funkyboy** [comment]
@mattheworiordan @ricardopereira I took a look and provided a few comments. 
BTW this has already been merged. 
In the future I think it's better if we aim for smaller PRs, because this was quite big :)

**Comment by mattheworiordan** [comment]
> In the future I think it's better if we aim for smaller PRs, because this was quite big :)

Agreed.  Although I think @ricardopereira was trying to fix all the test failures in one.  Still, I agree it's more likely to be reviewed and merged quicker if we don't have big PRs.

Thanks for reviewing @funkyboy and your contributions @ricardopereira

**Comment by ricardopereira** [comment]
@funkyboy Thank you for reviewing! I agree and sorry about that 😖 but sometimes _a man gotta do what a man gotta do_. Just kidding 😄. I usually do small PR's but this one was an exception. I needed a green CI and it's tedious to create a bunch of PR's to achieve it. The test suite was in a really bad state.

**Pull request merged**